# Introduction

WeRateDogs is a Twitter account that rates people's dogs with a humorous comment about the dog. WeRateDogs has over 4 million followers and has received international media coverage.

Your goal: wrangle WeRateDogs Twitter data to create interesting and trustworthy analyses and visualizations. The Twitter archive is great, but it only contains very basic tweet information. Additional gathering, then assessing and cleaning is required for "Wow!"-worthy analyses and visualizations.

## Gather

Here i will gather the required data from the different sources. My different sources are:
APi, .csv, .tsv.  
- the csv is loaded through read_csv method
- the tsv is downloaded programmatically
- the api data is queried and saved

In [170]:
import pandas as pd
# import rtsimple as rt
import requests
import tweepy.api as api
import json
from timeit import default_timer as timer
%run 'authorization.py' #importing macros from python file

In [171]:
#here load data form the csv and tsv

twitter_archive_df = pd.read_csv('twitter-archive-enhanced.csv') #
# image_predictions_1 = pd.read_csv('image-predictions-1.tsv', sep='\t')

In [172]:
#downloading file programmatically

url='https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response= requests.get(url)
with open('image-predictions.tsv','wb') as file:
        file.write(response.content)

In [173]:
image_predictions_df = pd.read_csv('image-predictions.tsv', sep='\t')  #what breed the dog is

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [175]:
#  twitter API authentication.
#have to use glob file here.

#trial wtih one tweet id.
consumer_key = API_KEY 
consumer_secret = API_SECRET 
access_token = ACCESS_TOKEN
access_secret = ACCESS_TOKEN_SECRET 

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify = True)

In [176]:
#data extraction from twitter api using tweet id from twitter_archive
fails_dict={}
            
lists =['666029285002620928','891815181378084864','892177421306343426']
start = timer()    
with open('tweet_json.txt', 'w') as outfile:
#     for tweet_id in twitter_archive.tweet_id[:100]:
    for tweet_id in twitter_archive.tweet_id:
        try:
            tweet = api.get_status(tweet_id, tweet_mode='extended')
#             print("Success")
            json.dump(tweet._json, outfile)
            outfile.write('\n')
        except tweepy.TweepError as e:
            print("Fail")
            fails_dict[tweet_id] = e
            pass
end = timer()
print(end - start)
print(fails_dict) #how to handle fails

Fail
Fail
Fail
Fail
Fail
Fail
Fail
Fail
Fail
Fail
Fail
Fail
Fail
Fail
Fail
Fail
Fail
Fail
Fail
Fail
Fail
Fail
Fail


Rate limit reached. Sleeping for: 332


Fail
Fail


Rate limit reached. Sleeping for: 335


2218.6700922269956
{888202515573088257: TweepError([{'code': 144, 'message': 'No status found with that ID.'}]), 873697596434513921: TweepError([{'code': 144, 'message': 'No status found with that ID.'}]), 872668790621863937: TweepError([{'code': 144, 'message': 'No status found with that ID.'}]), 872261713294495745: TweepError([{'code': 144, 'message': 'No status found with that ID.'}]), 869988702071779329: TweepError([{'code': 144, 'message': 'No status found with that ID.'}]), 866816280283807744: TweepError([{'code': 144, 'message': 'No status found with that ID.'}]), 861769973181624320: TweepError([{'code': 144, 'message': 'No status found with that ID.'}]), 856602993587888130: TweepError([{'code': 144, 'message': 'No status found with that ID.'}]), 851953902622658560: TweepError([{'code': 144, 'message': 'No status found with that ID.'}]), 845459076796616705: TweepError([{'code': 144, 'message': 'No status found with that ID.'}]), 844704788403113984: TweepError([{'code': 144, 'mes

In [163]:
print(fails_dict) 

{888202515573088257: TweepError([{'code': 144, 'message': 'No status found with that ID.'}]), 873697596434513921: TweepError([{'code': 144, 'message': 'No status found with that ID.'}]), 872668790621863937: TweepError([{'code': 144, 'message': 'No status found with that ID.'}]), 872261713294495745: TweepError([{'code': 144, 'message': 'No status found with that ID.'}]), 869988702071779329: TweepError([{'code': 144, 'message': 'No status found with that ID.'}]), 866816280283807744: TweepError([{'code': 144, 'message': 'No status found with that ID.'}]), 861769973181624320: TweepError([{'code': 144, 'message': 'No status found with that ID.'}]), 856602993587888130: TweepError([{'code': 144, 'message': 'No status found with that ID.'}]), 851953902622658560: TweepError([{'code': 144, 'message': 'No status found with that ID.'}]), 845459076796616705: TweepError([{'code': 144, 'message': 'No status found with that ID.'}]), 844704788403113984: TweepError([{'code': 144, 'message': 'No status f

Some tweet querins have failed. There are only two errors which are recurring:
- TweepError([{'code': 144, 'message': 'No status found with that ID.'}]) :Corresponds with HTTP 404. The requested Tweet ID is not found (if it existed, it was probably deleted)
- TweepError([{'code': 179, 'message': 'Sorry, you are not authorized to see this status.'}]) :Corresponds with HTTP 403. Thrown when a Tweet cannot be viewed by the authenticating user, usually due to the Tweet’s author having protected their Tweets.

In [164]:
# fails_dict
#how to handle fails
tweet_ids = twitter_archive.tweet_id.values
len(tweet_ids)

2356

In [177]:
#df to store additional info
count= 0
twitter_api = {}
twitter_api['tweet_id']=[]
twitter_api['retweet_count']=[]
twitter_api['favorite_count']=[]
with open('tweet_json.txt', 'r') as file:
#     print(file.readline())
#     j = json.loads(file.readline())
#     print(j['id_str'], j['retweet_count'],j['favorite_count'])
    for line in file.readlines():
        j = json.loads(line)
        count+=1
#         print(str(count)+': ' +j['id_str'],j['retweet_count'],j['favorite_count'])
        twitter_api['tweet_id'].append(j['id_str'])
        twitter_api['retweet_count'].append(j['retweet_count'])
        twitter_api['favorite_count'].append(j['favorite_count'])


In [178]:
twitter_api_df = pd.DataFrame(twitter_api)

## Assess

In [184]:
twitter_archive_df

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
...                  ...                    ...                  ...   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
0     2017-08-01 16:23:56 +0000   
1     2017-08-01 00:17:27 +0000   
2     2017-07-31 00:18:03 +0000   
3     2017-07-30 15:58:51 +0000   
4     2017-07-29 16:00:24 +0000   
...                         ...   
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
0     This is Phineas. He's a mystical boy. Only eve...                  NaN   
1     This is Tilly. She's just checking pup on you....                  NaN   
2     This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3     This is Darla. She commenced a snooze mid meal...                  NaN   
4     This is Franklin. He would like you to stop ca...                  NaN   
...                                                 ...                  ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
0                          NaN                        NaN   
1                          NaN                        NaN   
2                          NaN                        NaN   
3                          NaN                        NaN   
4                          NaN                        NaN   
...                        ...                        ...   
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
0     https://twitter.com/dog_rates/status/892420643...                13   
1     https://twitter.com/dog_rates/status/892177421...                13   
2     https://twitter.com/dog_rates/status/891815181...                12   
3     https://twitt

In [225]:
twitter_archive_df.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
655   791784077045166082                    NaN                  NaN   
1468  694206574471057408                    NaN                  NaN   
393   825876512159186944                    NaN                  NaN   
799   772826264096874500                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   

                      timestamp  \
655   2016-10-27 23:30:09 +0000   
1468  2016-02-01 17:11:59 +0000   
393   2017-01-30 01:21:19 +0000   
799   2016-09-05 15:58:34 +0000   
14    2017-07-24 17:02:04 +0000   

                                                 source  \
655   <a href="http://twitter.com/download/iphone" r...   
1468  <a href="http://twitter.com/download/iphone" r...   
393   <a href="http://twitter.com/download/iphone" r...   
799   <a href="http://twitter.com/download/iphone" r...   
14    <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
655   RT @dog_rates: I'm not sure what this dog is d...         6.820881e+17   
1468  "Martha come take a look at this. I'm so fed u...                  NaN   
393   This is Mo. No one will push him around in the...                  NaN   
799   Meet Roosevelt. He's preparing for takeoff. Ma...                  NaN   
14    This is Stuart. He's sporting his favorite fan...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
655               4.196984e+09  2015-12-30 06:37:25 +0000   
1468                       NaN                        NaN   
393                        NaN                        NaN   
799                        NaN                        NaN   
14                         NaN                        NaN   

                                          expanded_urls  rating_numerator  \
655   https://vine.co/v/iqMjlxULzbn,https://vine.co/...                12   
1468  https://twitter.com/dog_rates/status/694206574...                10   
393   https://twitter.com/dog_rates/status/825876512...                11   
799   https://twitter.com/dog_rates/status/772826264...                11   
14    https://twitter.com/dog_rates/status/889531135...                13   

      rating_denominator       name doggo floofer pupper  puppo  
655                   10       None  None    None   None   None  
1468                  10       None  None    None   None   None  
393                   10         Mo  None    None   None   None  
799                   10  Roosevelt  None    None   None   None  
14                    10     Stuart  None    None   None  puppo

In [182]:
twitter_archive_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [193]:
twitter_archive_df.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [223]:
twitter_archive_df.name.value_counts()

None       745
a           55
Charlie     12
Oliver      11
Lucy        11
          ... 
Dobby        1
Kona         1
Tommy        1
Geoff        1
Trip         1
Name: name, Length: 957, dtype: int64

In [211]:
twitter_archive_df.tweet_id.duplicated().any()

False

In [194]:
image_predictions_df

tweet_id                                          jpg_url  \
0     666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
...                  ...                                              ...   
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num                      p1   p1_conf  p1_dog                  p2  \
0           1  Welsh_springer_spaniel  0.465074    True              collie   
1           1                 redbone  0.506826    True  miniature_pinscher   
2           1         German_shepherd  0.596461    True            malinois   
3           1     Rhodesian_ridgeback  0.408143    True             redbone   
4           1      miniature_pinscher  0.560311    True          Rottweiler   
...       ...                     ...       ...     ...                 ...   
2070        2                  basset  0.555712    True    English_springer   
2071        1             paper_towel  0.170278   False  Labrador_retriever   
2072        1               Chihuahua  0.716012    True            malamute   
2073        1               Chihuahua  0.323581    True            Pekinese   
2074        1                  orange  0.097049   False               bagel   

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0     0.156665    True            Shetland_sheepdog  0.061428    True  
1     0.074192    True          Rhodesian_ridgeback  0.072010    True  
2     0.138584    True                   bloodhound  0.116197    True  
3     0.360687    True           miniature_pinscher  0.222752    True  
4     0.243682    True                     Doberman  0.154629    True  
...        ...     ...                          ...       ...     ...  
2070  0.225770    True  German_short-haired_pointer  0.175219    True  
2071  0.168086    True                      spatula  0.040836   False  
2072  0.078253    True                       kelpie  0.031379    True  
2073  0.090647    True                     papillon  0.068957    True  
2074  0.085851   False                       banana  0.076110   False  

[2075 rows x 12 columns]

In [195]:
image_predictions_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [196]:
image_predictions_df.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [212]:
image_predictions_df.tweet_id.duplicated().any()

False

In [197]:
twitter_api_df

tweet_id  retweet_count  favorite_count
0     892420643555336193           7532           35553
1     892177421306343426           5587           30751
2     891815181378084864           3702           23125
3     891689557279858688           7714           38855
4     891327558926688256           8316           37126
...                  ...            ...             ...
2326  666049248165822465             40              95
2327  666044226329800704            126             265
2328  666033412701032449             39             110
2329  666029285002620928             41             119
2330  666020888022790149            454            2368

[2331 rows x 3 columns]

In [198]:
twitter_api_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   tweet_id        2331 non-null   object
 1   retweet_count   2331 non-null   int64 
 2   favorite_count  2331 non-null   int64 
dtypes: int64(2), object(1)
memory usage: 54.8+ KB


In [199]:
twitter_api_df.describe()

retweet_count  favorite_count
count    2331.000000     2331.000000
mean     2648.198198     7429.762763
std      4478.608707    11538.132656
min         1.000000        0.000000
25%       537.000000     1292.000000
50%      1239.000000     3229.000000
75%      3074.500000     9096.000000
max     76087.000000   153457.000000

In [214]:
twitter_api_df.tweet_id.duplicated().any()

False

#### Quality Issues:
##### `twitter_archive_df` table
- Remove tweets that are replies or retweets
- Missing information (expanded_urls)
- Entry as `None` instead of NaN (dog stage variables, name)
- Erroneous datatypes (tweet_id, dog stage variables)
- Incorrect value of rating denominator (must be always = 10, not 0 or 170) **There are multiple dogs in pic, divide both numerator and denominator by no.of dogs in pic.**
- Invalid entries 'a' in name column.

##### `image_predictions_df` table
- Erroneous datatypes (tweet_id)
- Algorithm outcome is present.
- Inconsistency in font (p1,p2,p3)

##### `twitter_api_df` table


#### Tidiness Issues:
- One variable in four columns in `twitter_archive_df` table (doggo, floofer, puppo and pupper)
- Given tweet_id in `twitter_archive_df` table duplicated in `twitter_api_df` table form one observation unit.

##### `twitter_archive_df` table
- date and time are mixed in the timestamp column.
- Contact column in `patients` table should be split into phone number and email
- Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables

## Clean